examples/Raytracer_Benchmark/RaytraceBenchmark.hpp

```diff
@@ -0,0 +1,147 @@
+// SPDX-FileCopyrightText: 2020 CERN
+// SPDX-License-Identifier: Apache-2.0
+
+#include <VecGeom/base/Vector3D.h>
+#include <VecGeom/management/GeoManager.h>
+#include <VecGeom/navigation/NavStatePath.h>
+#include <VecGeom/base/Stopwatch.h>
+#include <VecGeom/management/CudaManager.h>
+
+#include "ArgParser.h"
+#include "examples/Raytracer_Benchmark/Raytracer.h"
+#include <CopCore/Global.h>
+#include <AdePT/BlockData.h>
+#include "kernels.h"
+#include "examples/Raytracer_Benchmark/LoopNavigator.h"
+
+#ifdef VECGEOM_GDML
+#include <VecGeom/gdml/Frontend.h>
+#endif
+
+namespace cuda {
+  struct RaytracerData_t;
+}
+
+void cudaWorld(cuda::RaytracerData_t *rtdata);
+
+template <copcore::BackendType backend>
+int runSimulation(const vecgeom::cxx::VPlacedVolume *world, int argc, char *argv[])
+{
+
+	// image size in pixels
+    OPTION_INT(px, 1840);
+    OPTION_INT(py, 512);
+
+    // RT model as in { kRTxray = 0, kRTspecular, kRTtransparent, kRTdiffuse };
+    OPTION_INT(model, 2);
+
+    // RT view as in { kRTVparallel = 0, kRTVperspective };
+    OPTION_INT(view, 1);
+
+    // zoom w.r.t to the default view mode
+    OPTION_DOUBLE(zoom, 3.5);
+
+    // Screen position in world coordinates
+    OPTION_DOUBLE(screenx, -5000);
+    OPTION_DOUBLE(screeny, 0);
+    OPTION_DOUBLE(screenz, 0);
+
+    // Up vector (no need to be normalized)
+    OPTION_DOUBLE(upx, 0);
+    OPTION_DOUBLE(upy, 1);
+    OPTION_DOUBLE(upz, 0);
+    vecgeom::Vector3D<double> up(upx, upy, upz);
+
+    // Light color, object color (no color per volume yet) - in RGBA chars compressed into an unsigned integer
+    OPTION_INT(bkgcol, 0xFF0000FF); // red
+    OPTION_INT(objcol, 0x0000FFFF); // blue
+    OPTION_INT(vdepth, 4);          // visible depth
+
+    copcore::Allocator<RaytracerData_t, backend> rayAlloc;
+    RaytracerData_t *rtdata = rayAlloc.allocate(1);
+  
+    rtdata->fScreenPos.Set(screenx, screeny, screenz);
+    rtdata->fUp.Set(upx, upy, upz);
+    rtdata->fZoom     = zoom;
+    rtdata->fModel    = (ERTmodel)model;
+    rtdata->fView     = (ERTView)view;
+    rtdata->fSize_px  = px;
+    rtdata->fSize_py  = py;
+    rtdata->fBkgColor = bkgcol;
+    rtdata->fObjColor = objcol;
+    rtdata->fVisDepth = vdepth;
+    rtdata->fMaxDepth = vecgeom::GeoManager::Instance().getMaxDepth();
+
+    Raytracer::InitializeModel((Raytracer::VPlacedVolumePtr_t)world, *rtdata);
+
+    if (backend == copcore::BackendType::CUDA) {
+
+  		cudaWorld((cuda::RaytracerData_t*)rtdata);
+
+    }
+    else {
+    	
+    	vecgeom::NavStateIndex vpstate;
+    	LoopNavigator::LocatePointIn(rtdata->fWorld, rtdata->fStart, vpstate, true);
+		rtdata->fVPstate = vpstate;
+
+    }
```

Comment:
This logic looks like a good candidate for a separate templated method InitRTdata<backend>